core/state/snapshot/dangling.go

```diff
@@ -0,0 +1,129 @@
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package snapshot
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// DanglingRange describes the range for detecting dangling storages.
+type DanglingRange struct {
+	db    ethdb.KeyValueStore // The database stores the snapshot data
+	start []byte              // The start of the key range
+	limit []byte              // The last of the key range
+
+	result   []common.Hash // The list of account hashes which have the dangling storages
+	duration time.Duration // Total time spent on the iteration
+}
+
+// NewDanglingRange initializes a dangling storage scanner and detects all the
+// dangling accounts out.
+func NewDanglingRange(db ethdb.KeyValueStore, start, limit []byte, report bool) *DanglingRange {
+	r := &DanglingRange{
+		db:    db,
+		start: start,
+		limit: limit,
+	}
+	r.result, r.duration = r.detect(report)
+
+	// Update metrics
+	snapDanglingStoragesCounter.Inc(int64(len(r.result)))
+	snapDanglingStoragesTimer.Update(r.duration)
+
+	if len(r.result) > 0 {
+		log.Warn("Detected dangling storages", "number", len(r.result), "start", hexutil.Encode(start), "limit", hexutil.Encode(limit), "elapsed", common.PrettyDuration(r.duration))
+	} else {
+		logger := log.Debug
+		if report {
+			logger = log.Info
+		}
+		logger("Verified snapshot storages", "start", hexutil.Encode(start), "limit", hexutil.Encode(limit), "elapsed", common.PrettyDuration(r.duration))
+	}
+	return r
+}
+
+// detect iterates the storage snapshot in the specified key range and
+// returns a list of account hash of the dangling storages. Note both
+// start and limit are included for iteration.
+func (r *DanglingRange) detect(report bool) ([]common.Hash, time.Duration) {
+	var (
+		checked    []byte
+		result     []common.Hash
+		start      = time.Now()
+		lastReport = time.Now()
+	)
+	iter := rawdb.NewKeyLengthIterator(r.db.NewIterator(rawdb.SnapshotStoragePrefix, r.start), len(rawdb.SnapshotStoragePrefix)+2*common.HashLength)
+	defer iter.Release()
+
+	for iter.Next() {
+		account := iter.Key()[len(rawdb.SnapshotStoragePrefix) : len(rawdb.SnapshotStoragePrefix)+common.HashLength]
```

Comment:
I know that the way you have written it is "more correct", but IMO it's easier to read on the form
```
account := iter.Key()[1:33] // trim 1-byte prefix, read 32-byte hash
```
Might be a minority opinion though ...